### PR TITLE
Add user-wide mirrord exec configuration

### PR DESCRIPTION
**Pull request by meowjesty**
## Summary

- Store user-wide mirrord configuration in the existing `~/.mirrord/data.json` file.
- Remember successful operator use per resolved Kubernetes context so later `mirrord exec` runs do not accidentally
  fall back to OSS mode.
- Support a user-wide default Kubernetes context for `mirrord exec`.
- Preserve explicit project configuration, environment variables, and CLI flags over user-wide defaults.

This PR is stacked on #4807, which is stacked on #4792. The CLI editing, import/export, and apply-to-all PRs are stacked above it.

Linear: https://linear.app/metalbear/issue/COR-1798/add-user-wide-mirrord-cli-configuration

- Translation:

Clanker is on point on this one. The original issue asked for a new config file, but we can reuse the `data.json` we
already have. This PR introduces the user config.

## Testing

- `cargo test --package mirrord user_config --target x86_64-unknown-linux-gnu`

### Quality Checklist

- [x] I have documented new code sufficiently
- [x] I have checked and updated the relevant existing docs in code, including removing outdated material
- [x] User-facing website documentation can follow the CLI review; command behavior is documented by `--help`
- [x] Existing website documentation is not affected
- [x] I have tested this change and know it succeeds and fails as expected
- [x] I have written unit tests
- [x] E2E coverage is provided by the complete COR-1798 stack
- [x] I have explained what this PR introduces and why, and linked to relevant context and the lower PR
- [x] I have introduced a short, clear and well-formatted changelog entry

### CODE REVIEW

File: changelog.d/+user-wide-exec-configuration.added.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Add user-wide configuration for remembering successful operator use per Kubernetes context and selecting a default
+Kubernetes context for `mirrord exec`.
```
Comment (aviramha) [inline changelog.d/+user-wide-exec-configuration.added.md:2]: It's not only for exec, it should be for all "mirrord operations" - up, exec, ls, etc.

File: mirrord/cli/src/config/user_config.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+/// Values accepted by `mirrord user-config unset`.
+#[derive(Args, Debug)]
+#[command(group(
+    ArgGroup::new("settings")
+        .required(true)
+        .multiple(true)
+        .args(["operator", "kube_context"])
+))]
```
Comment (aviramha) [inline mirrord/cli/src/config/user_config.rs]: I don't like this implementation. It makes us implement set/unset for each potential value, with distinct errors and all. We should define a schema and just let users edit based on schema so when we add new values we just update schema (and ofc apply config based on the new fields but that's for the logic part)

File: changelog.d/+user-wide-exec-configuration.added.md (hunk 1 above)
Comment (meowjesty) [reply changelog.d/+user-wide-exec-configuration.added.md:2]: I'm trying out stacked PRs, the PR for this comes here https://github.com/metalbear-co/mirrord/pull/4794

File: mirrord/cli/src/config/user_config.rs (hunk 2 above)
Comment (meowjesty) [reply mirrord/cli/src/config/user_config.rs]: So no more `--set operator`, `--set whatever`, instead the user just passes us a file or json, and then we merge it into their `user_config`?

Comment (aviramha) [reply mirrord/cli/src/config/user_config.rs]: no
still `--set whatever` but no specific implementation for each. like git config, or helm values. you can just pass abitrary --set as long as they match the schema

Comment (aviramha) [reply mirrord/cli/src/config/user_config.rs]: like now you ahve a "switch case" for each value, instead of just letting them set arbitrary values.
if I run
`git config set aasdasd.asdasds 1`
it just works - I don't want it to be exactly like that since I'd prefer we parse and make sure it is valid, but that's kinda hte interface instead of implementing *each* value set/get like Java..

Comment (meowjesty) [reply mirrord/cli/src/config/user_config.rs]: edit: nvm this, I can see how people/clankers would use this.

Comment (meowjesty) [reply mirrord/cli/src/config/user_config.rs]: https://github.com/metalbear-co/mirrord/pull/4804/changes#diff-0e4c1b0d55dd6c53632177cbd096ecbdb28c23b7844400f37bdbe3fba8a8cb11R1997-R2003

It's here, the test linked has an example.

clanker suggested using json pointer syntax for this type of command, since some configs may have `.` in the config itself, so json pointer would addresses this type of issue.

**Comment by meowjesty** [comment]
@greptile review again pls

File: mirrord/cli/src/connection.rs [hunk 3]
```diff
@@ -243,6 +245,14 @@ where
         .await
         .map_err(|error| CliError::friendlier_error_or_else(error, CliError::CreateAgentFailed))?;
 
+    if let Some(context) = api.kube_context()
+        && let Err(error) = UserConfig::remember_operator(context.to_owned()).await
```
Comment (0x00A5) [inline mirrord/cli/src/connection.rs:247]: should we also "forget_operator" if try connect fails or the operator discovery returns false?

**Review by 0x00A5 — COMMENTED** [review]
Looks good. Justa nit: can we organize `mirrord/cli/src/user_data.rs` and `mirrord/cli/src/mirrord_data.rs` to the same module? Like `mirrord/cli/src/data.rs`. And if user data need to be a sub module make it `mirrord/cli/src/data/user.rs`?

File: mirrord/cli/src/connection.rs (hunk 3 above)
Comment (meowjesty) [reply mirrord/cli/src/connection.rs:247]: I think we don't want to forget it, isn't the whole point that we mark this context as having the operator?

Comment (0x00A5) [reply mirrord/cli/src/connection.rs:247]: Yes, but in rare cases, a context may lose an operator installation (uninstalled, cluster resource purged, etc.). In those cases, storing operator `true` is no longer right, so we store the new state too (forget).

File: mirrord/cli/src/data/user_config.rs
Comment (0x00A5) [inline mirrord/cli/src/data/user_config.rs:1]: Oh? some new code got shifted to this PR from one latter in the stack?

File: mirrord/cli/src/connection.rs (hunk 3 above)
Comment (meowjesty) [reply mirrord/cli/src/connection.rs:247]: Don't disagree with you, but then, shouldn't the user manually go and `unset` the operator for that context?

Otherwise the user config feature kinda loses it's purpose (if it can automatically go back to allowing OSS mirrord).

File: mirrord/cli/src/data/user_config.rs
Comment (meowjesty) [reply mirrord/cli/src/data/user_config.rs:1]: I just told it to separate the things into modules under the new shared `data.rs` module.

File: mirrord/cli/src/data/global_config.rs [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        .await?;
+        Ok(())
+    }
+
+    /// Merges the global config into the project config without overriding explicit project, CLI,
+    /// or environment values.
+    ///
+    /// Only `operator` participates in this merge until another global setting is needed.
+    pub(crate) fn merge(&self, project_config: &mut LayerConfig) {
+        project_config.operator = project_config.operator.or(self.config.operator);
```
Comment (0x00A5) [inline mirrord/cli/src/data/global_config.rs:65]: I think this is not right. If `project_config.operator` is set, respect that value. If it is not set, use the one in global config.

File: mirrord/cli/src/data/global_config.rs [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+};
+
+use mirrord_config::{LayerConfig, config::ConfigContext};
+use serde_json::Value;
+
+use super::{default_path, update_at_path};
+
+/// "~/.mirrord/global-mirrord.json"
+static GLOBAL_CONFIG_PATH: LazyLock<PathBuf> =
+    LazyLock::new(|| default_path("global-mirrord.json"));
```
Comment (0x00A5) [inline mirrord/cli/src/data/global_config.rs]: super super nit: I think it's better to name this `~/.mirrord/mirrord.json`. The location of this `mirrord.json` determines its scope not its name, `global-mirrord.json`.

File: mirrord/cli/src/data/global_config.rs [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    /// Creates [`GlobalConfig`] from the default file path.
+    pub(crate) async fn from_default_path() -> io::Result<Self> {
+        Self::from_path(GLOBAL_CONFIG_PATH.as_path()).await
+    }
+
+    async fn from_path(path: &Path) -> io::Result<Self> {
+        // LayerConfig::resolve needs an existing file, so use the shared locked, atomic update
+        // path to initialize it and recover invalid JSON as an empty object. A raw map preserves
+        // arbitrary fields without persisting resolved LayerConfig defaults.
+        let _: BTreeMap<String, Value> = update_at_path(path, |_| {}).await?;
```
Comment (0x00A5) [inline mirrord/cli/src/data/global_config.rs]: We shouldn't use a `BTreeMap`. I think we should add a `RawGlobalConfig` type that exposes fields that we want global config to support. When read write a global config file via `update_at_path`, use `RawGlobalConfig` type instead of `BTreeMap`.

Same applies to the method `remember_operator_at_path`.  Replace `BTreeMap` with `RawGlobalConfig`, so we don't insert to the map and hardcode field name. Imagine one day we add `feature.network.incoming.http_fitler` in global config, that btreemap traversal can't look good 😆 .

We need a unit test to make sure this path always pass, `RawGlobalConfig` --> a file -->  `LayerConfig`, to guard breaking change in `LayerConfig`, (even though it is not a concern right now since we have only one field).

File: mirrord/cli/src/data/global_config.rs (hunk 4 above)
Comment (0x00A5) [reply mirrord/cli/src/data/global_config.rs:65]: nit: `a.merge(b)` means fold `b` into `a`. This should be `a.apply_to(b)`

File: mirrord/cli/src/main.rs [hunk 7]
```diff
@@ -793,7 +793,13 @@ async fn exec(
...
     cfg_context = apply_test_env_overrides(cfg_context);
 
-    let (config_file_path, mut config) = util::resolve_config(&mut cfg_context)?;
+    let global_config = GlobalConfig::from_default_path()
+        .await
+        .inspect_err(|fail| trace!(?fail, "Failed initializing global mirrord config"))
+        .unwrap_or_default();
+
+    let (config_file_path, mut config) =
+        util::resolve_config_with_global_config(&mut cfg_context, &global_config)?;
```
Comment (0x00A5) [inline mirrord/cli/src/main.rs:802]: I think we don't want this method `resolve_config_with_global_config`. Applying global config only affects the config, and not the `config_file_path`, so no point wrapping `resolve_config` function.


```rust
    let (config_file_path, mut config) = util::resolve_config(&mut cfg_context)?;
    let global_config = GlobalConfig::from_default_path()
        .await
        .inspect_err(|fail| trace!(?fail, "Failed initializing global mirrord config"))
        .unwrap_or_default();
    global_config.apply_to(&mut config);
```

File: mirrord/cli/src/data/global_config.rs (hunk 4 above)
Comment (meowjesty) [reply mirrord/cli/src/data/global_config.rs:65]: e1adbcee7c9c8a502a4fbc8c07b7de6c86cc3f23

> I think this is not right. If project_config.operator is set, respect that value. If it is not set, use the one in global config.

I'm not sure what you mean here, I tested this out with a local config that has `operator: false`, it overrides the global `operator: true`.

I've also tested out no operator in cluster (by returning `None` instead of the api), and global `operator: true` fails (same if we set the local config to that).

Isn't the thing you're saying, what we're doing? We check the `project_config.operator`, and if it's not set we use the `self.config.operator` (global config)?

File: mirrord/cli/src/data/global_config.rs (hunk 5 above)
Comment (meowjesty) [reply mirrord/cli/src/data/global_config.rs]: 24569ef71c2067152b6e099a427db57226c49702

File: mirrord/cli/src/main.rs (hunk 7 above)
Comment (meowjesty) [reply mirrord/cli/src/main.rs:802]: This PR adds it only for exec, but the other commands also need to handle this whole thing.

File: mirrord/cli/src/data/global_config.rs (hunk 6 above)
Comment (meowjesty) [reply mirrord/cli/src/data/global_config.rs]: Can we put this as a sub-issue/future task? Right now doing anything other than just having a json that takes whatever we throw at it seems like so much more effort than what's needed?

This could even be something like a `serde_json::Value` (or `Object`, which is just the `BTreeMap` lol), right?

> so we don't insert to the map and hardcode field name

I don't get what you meant by this. Hardcode the name like `feature.fs.read_only`, something like that? What is the problem with this?

Comment (0x00A5) [reply mirrord/cli/src/data/global_config.rs]: You can have a raw config type that is like, that only has the fields we want the global config to support:
```
struct RawGlobalConfig {
    operator: Option<bool>,
}
```

Use this type for file read from or write to the file `~/.mirrord/mirorrd.json`.

With `BTreeMap<String, Value>`, mutating `feature.fs.read_only` is either, the following that could panic:
```
if let Some(feature) = map.get_mut("feature") {
    feature["fs"]["read_only"] = json!(true);
}
```

or something like

```
let updated = map
    .get_mut("feature")
    .and_then(|v| v.as_object_mut())
    .and_then(|obj| obj.get_mut("fs"))
    .and_then(|v| v.as_object_mut())
    .and_then(|obj| obj.get_mut("read_only"));

if let Some(v) = updated {
    *v = serde_json::json!(true);
}
```

But with `RawGlobalConfig`, it is `raw_global_config.feature.fs.read_only = true`.

File: mirrord/cli/src/data/global_config.rs (hunk 4 above)
Comment (0x00A5) [reply mirrord/cli/src/data/global_config.rs:65]: Oh I misread, the code is right. `global.operator` only overrides when `project_config.operator` is None already.

File: mirrord/cli/src/data.rs [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+
+/// Creates a missing JSON document as an empty object without parsing or rewriting existing data.
+pub(crate) async fn initialize_empty_json_at_path(path: &Path) -> io::Result<()> {
+    let path = path.to_owned();
+    task::spawn_blocking(move || {
+        if let Some(parent) = path.parent() {
+            fs::create_dir_all(parent)?;
+        }
+
+        let lock_path = path.with_extension("lock");
```
Comment (0x00A5) [inline mirrord/cli/src/data.rs:140]: Should we only create lock file in the `Err(error) if error.kind() == io::ErrorKind::NotFound` match branch below?

File: mirrord/cli/src/data.rs [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+            }
+            Some(Err(error)) => {
+                return Err(io::Error::new(io::ErrorKind::InvalidData, error));
+            }
+            None => T::default(),
+        };
+
+        update(&mut data);
+
+        let contents = serde_json::to_vec(&data).map_err(io::Error::other)?;
```
Comment (0x00A5) [inline mirrord/cli/src/data.rs]: I think we should print the prettified JSON value to the file. Right now if the user edits the file (put some extra config as global settings), we replace the file content with compacted JSON.

For example,
```
{
	"operator": true,
	"experimental": {
		"go_asmcgocall": false
	}
}
```

after running a session, becomes:
```
{"operator":true,"experimental":{"go_asmcgocall":false}}
```

**Review by 0x00A5 — APPROVED** [review]
